streaming/base/util.py

```diff
@@ -196,3 +204,138 @@ def get_import_exception_message(package_name: str, extra_deps: str) -> str:
     return f'Streaming was installed without {package_name} support. ' + \
             f'To use {package_name} related packages with Streaming, run ' + \
             f'`pip install \'mosaicml-streaming[{package_name}]\'`.'
+
+
+def merge_index(folder_urls: List[Union[str, Tuple[str, str]]],
+                out: Union[str, Tuple[str, str]],
+                *,
+                keep_local: bool = True,
+                overwrite: bool = True,
+                download_timeout: int = 100) -> int:
+    """Merge index.json from a list of remote or local directories.
+
+    Args:
+        folder_urls (Iterable): folders that contain index.json for the partition
+            each element can take the form of a single path string or a tuple string
+
+            for each  url in folder_urls, if url is
+                1. tuple (local, remote): check if local is accessible.
+                    -> Yes: use local index to merge
+                    -> No:  download from remote first, then merge
+                2. str (local path): use local path to merge.
+                    raise FileNotFoundError if any local index is not accessible
+                3. str (remote url): download to a temp directory first, then merge
+
+        out (Union[str, Tuple[str, str]]): path to put the merged index file
+        keep_local (bool): Keep local copy of the merged index file. Defaults to ``True``
+        overwrite (bool): Overwrite merged index file in out if there exists one.Defaults to ``True``
+        download_timeout (int): The allowed time for downloading each json file
+            defaults to 60, same as streaming.download_file
+
+    Returns:
+        int: count of files downloaded during function call
+    """
+    # Import here to avoid circular import error
+    from streaming.base.storage.download import download_file
+    from streaming.base.storage.upload import CloudUploader
+
+    if not folder_urls:
+        logger.warning('No partitions exist, no index merged')
+        return 0
+
+    # This is the index json file name, e.g., it is index.json as of 0.6.0
+    index_basename = get_index_basename()
+
+    cu = CloudUploader.get(out, keep_local=True, exist_ok=True)
+    if os.path.exists(os.path.join(cu.local, index_basename)) and overwrite:
+        logger.warning('Merged index already exists locally. no index merged if overwrite=False')
+        return 0
+
+    # Remove '/' from right, so os.path.basename gives relative path to each folder
+    urls = []
+    for url in folder_urls:
+        if type(url) is str:
```

Comment:
nit: this also works but I would recommend `isinstance(url, str)` since it is faster than `type` check.

Reply:
Done.